### ISSUE TITLE
Samplewise Centering is implemented wrong

### ISSUE THREAD

**Issue by StefanoD** [keras-team/keras#6718]
I stumbled over this bug, because the hit hit-rate of my dataset dropped from > 97% to ~92% when I switched from my own standardization code to the Keras code.
The [current](https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py#L495) code calculates the mean _through_ the channels:

```python
x -= np.mean(x, axis=img_channel_axis, keepdims=True)
```
In my opinion this is not good. If you have three channels, you calculate the mean of three pixels over all channels. If you have only one channel, the mean is the pixel value itself and the std is 0.

### What is the right way?

Flatten the image _per channel_ and calculate the mean and then subtract the mean from the respective channel.

### What does Tensorflow?
As far as I can tell, it calculates the mean over all channels. It flattens the complete image and then calculates the mean. This is better then Keras approach, but still worse then mine, because the channels don't have to be color images. In my case, these three channels are recordings of the same object, but from different perspective. So it is better to calculate the mean channel-wise. This would work in all use cases well.
Here the code from [Tensorflow](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_impl.py#L817):

```python
image_mean = math_ops.reduce_mean(image)
```

And [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_ops.py#L1311) the documentation when axis is None:

  ```python
  # 'x' is [[1., 1.]
  #         [2., 2.]]
  tf.reduce_mean(x) ==> 1.5
  ```

**Comment by alexhernandezgarcia** [comment]
I agree, the current standardization is wrong. It computes and subtracts the RGB mean of **each pixel**.

It should compute and subtract either the mean of **all** pixels:
`x -= np.mean(x, axis=(0, 1, 2), keepdims=True)`

or the mean of all the pixels in each channel independently:
`x -= np.mean(x, axis=(0, 1), keepdims=True)`

Same applies for the division by the standard deviation!

**Comment by StefanoD** [comment]
Ok, guys. I give up. [PR](https://github.com/fchollet/keras/pull/6736) hasn't been merged till now...

**Comment by harshini-gadige** [comment]
Is this still an issue ?
